### DISCUSSION TITLE
Is it possible to convert a regular vector to a `ComponentArray` with the specified structure?

### DISCUSSION THREAD

**Discussion by ForceBru** [jonniedie/ComponentArrays.jl#159] [Q&A] [answered]
The README shows that a `ComponentArray` can be indexed like a vector and converted to a vector as well:

https://github.com/jonniedie/ComponentArrays.jl/blob/f3a1b6d9e6d1549f7513579632d66597067eb93b/README.md?plain=1#L44-L58

Is it possible to _reverse this operation_: convert a regular vector like `[1, 2]` to, say, a `ComponentArray(a=1, b=2)` with the given fields and their types? I want something like this to convert a vector to model parameters:

```
function loss(params_vec::AbstractVector)
    params = toComponentArray(params_vec, example=ComponentArray(a=1, b=2, c=[3, 4]))
    @. params.a + params.b + params.c
end

optimize(loss, [1, 2, 3, 4])
```

I know that one can pass the `ComponentArray` as the parameter because `ComponentVector <: AbstractVector`, but that won't work if the optimizer always calls `loss` with regular vectors for some reason.

Basically, I'd like to have a function like `toComponentArray` which receives the source vector and the expected structure of the `ComponentArray` and returns the `ComponentArray` that contains the data from the vector. Is this possible?

**Comment by ForceBru** [comment, marked as answer]
Looks like this is done by [`ComponentArrays.Axis`](https://jonniedie.github.io/ComponentArrays.jl/dev/api/#ComponentArrays.Axis):

https://github.com/jonniedie/ComponentArrays.jl/blob/f3a1b6d9e6d1549f7513579632d66597067eb93b/src/axis.jl#L20-L27

---

Yep, this seems to work and even has similar interface to what I envisioned:

```
julia> example = ComponentArray(a=1, b=5, c=[600,1000])
ComponentVector{Int64}(a = 1, b = 5, c = [600, 1000])

julia> ax, = getaxes(example)
(Axis(a = 1, b = 2, c = 3:4),)

julia> ComponentArray([2,3,9,8], ax)
ComponentVector{Int64}(a = 2, b = 3, c = [9, 8])
```

**Reply by ForceBru** [reply]
Thanks, me!

**Reply by NilsWildt** [reply]
Thank you ;)
